File: src/backend/internal/repository/postgres/issue.go
package postgres

import (
	"context"
	"time"

	"github.com/Roongkun/software-eng-ii/internal/model"
	"github.com/google/uuid"
	"github.com/uptrace/bun"
)

type IssueDB struct {
	*BaseDB[model.Issue]
}

func NewIssueDB(db *bun.DB) *IssueDB {
	type T = model.Issue

	return &IssueDB{
		BaseDB: NewBaseDB[T](db),
	}
}

func (i *IssueDB) FindIssuesWithFilter(ctx context.Context, filter model.IssueFilter) ([]*model.Issue, error) {
	var issues []*model.Issue
	query := i.db.NewSelect().Model(&issues)

	if filter.Subject != nil {
		query = query.Where("subject = ?", *filter.Subject)
	}
	if filter.CreatedAt != nil {
		query = query.Where("created_at = ?", *filter.CreatedAt)
	}
	if filter.DueDate != nil {
		query = query.Where("due_date = ?", *filter.DueDate)
	}
	if filter.ReporterId != nil {
		query = query.Where("reporter_id = ?", uuid.MustParse(*filter.ReporterId))
	}
	if filter.Status != nil {
		query = query.Where("status = ?", *filter.Status)
	}

	if err := query.Scan(ctx, &issues); err != nil {
		return nil, err
	}

	return issues, nil
}

func (i *IssueDB) GetIssueHeaderMetadata(ctx context.Context) (*model.IssueHeaderMetadata, error) {
	var issue model.Issue
	var result model.IssueHeaderMetadata

	count, err := i.db.NewSelect().Model(&issue).Where("status = ?", model.IssueOpenStatus).Count(ctx)
	if err != nil {
		return nil, err
	}
	result.PendingTickets = count

	date := time.Now().Format("2006-01-02")
	count, err = i.db.NewSelect().Model(&issue).Where("DATE(created_at) = ?", date).Count(ctx)
	if err != nil {
		return nil, err
	}
	result.TicketsToday = count

	count, err = i.db.NewSelect().Model(&issue).Where("DATE(due_date) = ?", date).Count(ctx)
	if err != nil {
		return nil, err
	}
	result.TicketsDueToday = count

	count, err = i.db.NewSelect().Model(&issue).Where("status = ?", model.IssueClosedStatus).Count(ctx)
	if err != nil {
		return nil, err
	}
	result.ClosedTickets = count

	return &result, nil
}

File: src/frontend/src/types/issue.ts
import { User } from "next-auth";
import { Booking } from "./booking";

export enum IssueStatus {
  Open = "OPEN",
  Closed = "CLOSED",
}

export enum IssueSubject {
  Refund = "REFUND",
  Technical = "TECHNICAL",
}

export interface Issue {
  id: string;
  reporterId: string;
  reporter: User;
  bookingId?: string;
  booking?: Booking;
  status: string;
  subject: string;
  due_date: Date;
  description: string;
  created_at: Date;
}

export interface IssueInput {
  description?: string;
}

export interface IssueFilter {
  reporterId?: string;
  status?: string;
  dueDate?: Date;
  createdAt?: Date;
  subject?: string;
}

export interface IssueHeaderMetadata {
  pending_tickets: number;
  tickets_today: number;
  tickets_dueToday: number;
  closed_tickets: number;
}
